sympy/vector/functions.py

```diff
@@ -196,24 +216,68 @@ def gradient(scalar, coord_sys):
     scalar : SymPy Expr
         The scalar field to compute the gradient of
 
-    coord_sys : CoordSysCartesian
-        The coordinate system to calculate the gradient in
-
     Examples
     ========
 
-    >>> from sympy.vector import CoordSysCartesian, gradient
-    >>> R = CoordSysCartesian('R')
+    >>> from sympy.vector import CoordSystem3D, gradient
+    >>> R = CoordSystem3D('R')
     >>> s1 = R.x*R.y*R.z
-    >>> gradient(s1, R)
+    >>> gradient(s1)
     R.y*R.z*R.i + R.x*R.z*R.j + R.x*R.y*R.k
     >>> s2 = 5*R.x**2*R.z
-    >>> gradient(s2, R)
+    >>> gradient(s2)
     10*R.x*R.z*R.i + 5*R.x**2*R.k
 
     """
+    # Check input
+    if isinstance(scalar, Zero) or scalar == 0:
+        return Vector.zero
+
+    if not issubclass(type(scalar), Expr) or scalar.is_Vector:
+        raise TypeError("Your input should be a scalar field.")
+
+    for atom in scalar.atoms():
+        if isinstance(atom, BaseScalar):
+            system = atom.system
+            break
```

Comment:
One simple example, suppose you want to calculate the gradient (in 3D) of `x*r`, you should calculate either:

* the gradient of `x*sqrt(x**2 + y**2 + z**2)`
* the gradient of `r**2*sin(theta)*cos(phi)`

The results are expressed in different coordinates, but should agree.

A simpler example, calculate the gradient of `x*xshift` where `xshift = x + delta` is a scalar field of a shifted coordinate system. The result is `(2*x + delta)*v_x` or `(2*xshift - delta)*v_xshift`.

This PR is already too large to handle all these cases. Maybe it's better not to address this point for now.